### PR TITLE
Jumper T20

### PR DESCRIPTION
**Pull request by 3djc**
This brings support for Jumper T20

6 function switches
6 pot/sliders
10 switches
8 trim channels

This radio does not feature an SD Card reader, but uses a sd chip (not card) on motherboard itself (no more crap, fake sd)

T20英文详情页切片_01

### CODE REVIEW

File: radio/util/hw_defs/switch_config.py [hunk 1]
```diff
@@ -126,6 +126,20 @@
...
+        "SE": { "default": "2POS",   "display": [ 0, 2 ] },
+        "SG": { "default": "TOGGLE",   "display": [ 0, 3 ] },
+        "SI": { "default": "TOGGLE",   "display": [ 0, 4 ] },
+        # right side
+        "SB": { "default": "2POS",   "display": [ 1, 0 ] },
+        "SD": { "default": "3POS",   "display": [ 1, 1 ] },
+        "SF": { "default": "2POS",   "display": [ 1, 2 ] },
+        "SH": { "default": "TOGGLE",   "display": [ 1, 3 ] },
+        "SJ": { "default": "TOGGLE",   "display": [ 1, 4 ] }
+    },
```
Comment (pfeerick) [inline radio/util/hw_defs/switch_config.py:142]: Not sure why this got duplicated

**Comment by pfeerick** [comment]
Can I get a translation for the following please... nice simple one this time. 😃 It is literally just that phrase - the TPro and T20 have the concept of "function switches" rather than the traditional 6POS switch. 

```
#define TR_FUNCTION_SWITCHES "Function switches"
```

- [ ] @zyren CN / TW
- [ ] @Eldenroot CZ
- [ ] @HThuren DA
- [ ] @ParkerEde / @TheIsotopes DE
- [ ] @Pat6874 FR
- [ ] @offer-shmuely HE
- [x] robustini IT - sourced from prior #3738
- [ ] @ToshihiroMakuuchi JP
- [ ] @ulfhedlund SE
- [ ] @ajjjjjjjj PL

**Comment by Pat6874** [comment]
#define TR_FUNCTION_SWITCHES "Function switches"

FR: 
#define TR_FUNCTION_SWITCHES "Fonction inters"

Le 2 juil. 2023, 09:31, à 09:31, Peter Feerick ***@***.***> a écrit:

**Comment by HThuren** [comment]
DA
#define TR_FUNCTION_SWITCHES "Funktion kontakter"

File: radio/util/hw_defs/switch_config.py (hunk 1 above)
Comment (3djc) [reply radio/util/hw_defs/switch_config.py:142]: Clion is to blame there I think !

**Comment by 3djc** [comment]
@pfeerick beware that similar string with different twist (casing and BTN) exist and where untranslated before : TR_MENU_FSWITCH and TR_FS_BTN

**Comment by pfeerick** [comment]
@Pat6874 @HThuren Sorry, there was a last minute change, it's now going to be "Customizable switches", so please see the changes to https://github.com/EdgeTX/edgetx/pull/3726#issuecomment-1616421976

**Comment by Eldenroot** [comment]
CZ:

#define TR_FUNCTION_SWITCHES "Nastavitelné přepínače"
#define TR_MENU_FSWITCH "NASTAVITELNÉ PŘEPÍNAČE"
#define TR_FS_BTN BUTTON(TR("Vlastní Sw", "Nastavitelné přepínače"))

**Comment by Pat6874** [comment]
Customizable switches

FR:
Inters paramétrables 

Le 2 juil. 2023, 11:30, à 11:30, Peter Feerick ***@***.***> a écrit:

**Comment by offer-shmuely** [comment]
HE
#define TR_FUNCTION_SWITCHES "מפסקי פונקציות"

**Comment by HThuren** [comment]
DA (all together)
#define TR_FUNCTION_SWITCHES "Funktion kontakter"
#define TR_MENU_FSWITCH           "Tilpassede kontakter"
#define TR_FS_BTN                          BUTTON(TR("Tilpas. kon.", TR_MENU_FSWITCH))

ok ?

**Comment by ulfhedlund** [comment]
SE: #define TR_FUNCTION_SWITCHES "Funktionsbrytare"Den 2 juli 2023 09:31 skrev Peter Feerick ***@***.***>:
Can I get a translation for the following please... nice simple one this time. 😃 It is literally just that phrase - the TPro and T20 have the concept of "function switches" rather than the traditional 6POS switch.
#define TR_FUNCTION_SWITCHES "Function switches"

 @zyren CN / TW @Eldenroot CZ @HThuren DA @ParkerEde / @TheIsotopes DE @Pat6874 FR @offer-shmuely HE robustini IT - sourced from prior #3738 @ToshihiroMakuuchi JP @ulfhedlund SE @ajjjjjjjj PL

**Comment by ParkerEde** [comment]
DE
#define TR_FUNCTION_SWITCHES "Anpassbare Schalter"
#define TR_MENU_FSWITCH "Anpassbare Schalter"
#define TR_FS_BTN BUTTON(TR("AnpSchalt", "Anpassbare Schalter"))

**Comment by zyren** [comment]
CN
```c
#define TR_FUNCTION_SWITCHES    "可自定义开关"
#define TR_MENU_FSWITCH         "可自定义开关"
#define TR_FS_BTN BUTTON        (TR("自定义开关", "可自定义开关"))
```
TW 
```c
#define TR_FUNCTION_SWITCHES    "可自定義開關"
#define TR_MENU_FSWITCH         "可自定義開關"
#define TR_FS_BTN BUTTON        (TR("自定義開關", "可自定義開關"))
```

**Comment by pfeerick** [comment]
@HThuren This was the edit I mentioned earlier ;)

> edit: Last minute change... we're going to call them "Customizable switches" instead to better reflect what they are.

**Comment by offer-shmuely** [comment]
HE
#define TR_FUNCTION_SWITCHES "מפסקים בהתאמה אישית"

**Comment by ajjjjjjjj** [comment]
PL
```
#define TR_FUNCTION_SWITCHES "Ustawiane przełączniki"
#define TR_MENU_FSWITCH "USTAWIANE PRZEŁĄCZNIKI"
#define TR_FS_BTN BUTTON(TR("Ust. łącz.", "Ustawiane przełączniki"))
```

**Comment by pfeerick** [comment]
@offer-shmuely Can I also get a more abbreviated form for "Custom sw"

@ulfhedlund Does this work for SE? (from "Custom sw", "Customizable switches")
```cpp
#define TR_FS_BTN                       BUTTON(TR("Anpassad sw", "Anpassningsbara brytare"))
```

@HThuren What about this for DA? (from "Custom sw", "Customizable switches")
```cpp
#define TR_FS_BTN                       BUTTON(TR("Tilpasset sw", "Kontakter der kan tilpasses"))
```

**Comment by robustini** [comment]
IT
#define TR_FUNCTION_SWITCHES "Interruttori personalizzabili"
#define TR_MENU_FSWITCH "INTERRUTTORI PERSONALIZZABILI"
#define TR_FS_BTN BUTTON(TR("Int. pers.", "Interruttori personalizzabili"))

**Comment by zandorsp** [comment]
PT 
> ```
> #define TR_FUNCTION_SWITCHES "Botão customizável"
> #define TR_MENU_FSWITCH "BOTÃO CUSTOMIZÁVEL"
> #define TR_FS_BTN BUTTON(TR("Bt Custom", "Botão customizável"))
> ```

**Comment by ulfhedlund** [comment]
SE
#define TR_FUNCTION_SWITCHES "Anpassningsbara brytare"
#define TR_MENU_FSWITCH "ANPASSNINGSBARA BRYTARE"
#define TR_FS_BTN BUTTON(TR("Anp. bryt.", "Anpassningsbara brytare"))

Too long?

**Comment by pfeerick** [comment]
@ulfhedlund Shorter than what I had, so hopefully *short enough* ;)

@zandorsp Do you mind being pinged for Portuguese translations in the future?

**Comment by HThuren** [comment]
> @ulfhedlund Shorter than what I had, so hopefully _short enough_ ;)
> 

10 or 11 char for the short / BW form ?

**Comment by pfeerick** [comment]
It gets shown here

image

so probably 10 for no clipping, 11 will probably clip the trailing `]` depending on the letters used.

**Comment by zandorsp** [comment]
> @zandorsp Do you mind being pinged for Portuguese translations in the future?

No... You can add me

**Comment by ToshihiroMakuuchi** [comment]
JP
#define TR_FUNCTION_SWITCHES "スイッチ カスタマイズ"
#define TR_MENU_FSWITCH "スイッチ カスタマイズ"
#define TR_FS_BTN BUTTON(TR("Custom sw", "スイッチ カスタマイズ"))

**Comment by derFrickler** [comment]
is there already a Companion build that supports the T20?

**Comment by 3djc** [comment]
Not yet

**Comment by derFrickler** [comment]
Hi, i tried to connect my old 4in1 Multimodule to the external Port of the T20. I am able to bind and stear models, but i do not get any feedback (telemetry, module version etc..) from the module. 

Am i doing something wrong or is this an issue with the T20/EdgeTX?  Normally i only need one pin/cable to the module for TX and TX right?

**Comment by pfeerick** [comment]
You should only need the one five-pin cable, yes. I've just tried with my T20, and RM 4IN1 module, and it worked fine. What make/model is your MPM?

**Comment by derFrickler** [comment]
Thanks, for the fast answer! 
Well, my goal is to have a [spare internal 4in1 from the T-Lite](https://www.banggood.com/de/Original-JumperRC-P10309-Multiple-Protocol-TX-Module-Replacement-Parts-for-Jumper-T-lite-Transmitter-p-1872557.html) inside the T20. 
I tested with that module and my fullsize Jumper 4in1. 
Guess i found the problem now, i was assuming that RX and TX go over one Cable as on the ELRS modules, but the 4in1 seems to use a separate wire for the telemetry.....makes sense as the T-Lite board has two solder pads....

do you have the pinout of the five-pin cable on hand? cant find it anywhere...

**Comment by pfeerick** [comment]
I don't know what it connects to internally, but it is literally a 1 to 1 mapping for the rightmost five pins (leftmost if looking at the module from the back) of the nano module bay... IIRC it's the same as the TPro nano module adapter (not the T-Lite). 

The only pinouts I've come across for the nano bay so far are these (I did have one more at one point, for the two UART pins used by ACCESS modules, but neglected to keep that filed away properly)... but in the case of the external MPM, it needs both the top (PPM) and bottom (SPort) pins for normal JR micro bay (and obviously two more for power). 

Pasted image 20230323160033
Pasted image 20230323160043

**Comment by pfeerick** [comment]
A quick continuity check of the BetaFPV JR micro adapter suggests (looking at the module bay)

| 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 |
| --- | --- | --- | --- | --- | --- | --- | --- |
| --- | --- | --- |PPM|HB|VBAT|GND|SPort|

**Comment by derFrickler** [comment]
YES, got it working! 
I a was confused as the "top" PPM Pin is the TX to the multimodule and the "bottom" CRFS S-Port is the RX from the multimodule.   I will do some pics when putting everything together. Shall i post them here or in a different "Ticked"

**Comment by pfeerick** [comment]
Excellent. Maybe throw it up on the discussions are under "Show and Tell" ?
https://github.com/EdgeTX/edgetx/discussions/categories/show-and-tell

**Comment by derFrickler** [comment]
As its more related to the Jumper T20, i think it does not fit in the EdgeTX dev. 
I posted some pics in the T20 Thread on rcgroups: 
https://www.rcgroups.com/forums/showpost.php?p=51314271&postcount=30

**Comment by derFrickler** [comment]
Is there a way to use the 6-pos Switch on the T20 to set one channel to 6 different positions? 
I saw that there is a 6P Input source on the Jumper T18 that does that, but i can't find it on the T20.

**Comment by 3djc** [comment]
There is no 6pos on t20, those are 6 distinct switches, that you can group to your liking. You can achieve what you want using mixer line with 6 entries for example

**Comment by derFrickler** [comment]
Thanks, did that.

**Comment by chakflying** [comment]
That seems overly-complicated and I have no idea how to do so. It would be much better if an N-position virtual switch is automatically created when you assign N switches to a group.

**Comment by pfeerick** [comment]
Have a look at https://github.com/EdgeTX/edgetx/issues/4189#issuecomment-1757186041

**Pull request merged**